util/syspolicy/setting/snapshot.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package setting
+
+import (
+	"slices"
+	"strings"
+
+	xmaps "golang.org/x/exp/maps"
+	"tailscale.com/util/deephash"
+)
+
+// Snapshot is an immutable collection of ([Key], [RawItem]) pairs, representing
+// a set of policy settings applied at a specific moment in time.
+// A nil pointer to [Snapshot] is valid.
+type Snapshot struct {
+	m       map[Key]RawItem
+	sig     deephash.Sum // of m
+	summary Summary
+}
+
+// NewSnapshot returns a new [Snapshot] with the specified items and options.
+func NewSnapshot(items map[Key]RawItem, opts ...SummaryOption) *Snapshot {
+	return &Snapshot{m: items, sig: deephash.Hash(&items), summary: SummaryWith(opts...)}
+}
+
+type keyItemPair struct {
+	Key  Key
+	Item RawItem
+}
+
+// All returns an iterator over [[Key], [RawItem]] key-value pairs in b. The
+// iteration order is not specified and is not guaranteed to be the same from
+// one call to the next.
+func (s *Snapshot) All() []keyItemPair {
+	if s == nil {
+		return nil
+	}
+	// TODO(nickkhyl): return iter.Seq2[[Key], [RawItem]] in Go 1.23,
+	// and remove [keyItemPair].
+	items := make([]keyItemPair, 0, len(s.m))
+	for k, i := range s.m {
+		items = append(items, keyItemPair{k, i})
+	}
+	return items
+}
+
+// Get returns the value of the policy setting with the specified key
+// or nil if it does not exist or could not be read.
+func (s *Snapshot) Get(k Key) any {
+	v, _ := s.GetErr(k)
+	return v
+}
+
+// GetErr returns the value of the policy setting with the specified key,
+// [ErrNotConfigured] if it does not exist, or an error returned by
+// the policy Store if the policy setting could not be read.
+func (s *Snapshot) GetErr(k Key) (any, error) {
+	if s != nil {
+		if s, ok := s.m[k]; ok {
+			return s.Value(), s.Error()
```

Comment:
Yeah, I think passing the actual policy store's error is better than a text recreation of it. A wrapped error would be fine too if it adds value.

Reply:
The `error` interface cannot be sent on the wire.

With an exception of `ErrNotConfigured` (and the legacy `ErrNoSuchKey`), which are handled differently, I think error type information is unimportant at the call sites.
But I think we'd like to preserve error messages for logging and similar purposes, and it should be JSON-marshalable (and unmarshalable) so that we can send it from the backend to the client.

If this happens to be incorrect, and we will need to distinguish between more error types when sending policy settings on the wire, we could encode that type information in the `Error` string (e.g., as an error code preceding the text) and make `Error` implement the `Unwrap` method. However, we can't do that with standard errors, and I think having marshalable errors is more important in syspolicy.